gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,393 @@
+public final class BigLakeCatalog extends BaseMetastoreCatalog
+    implements Closeable, SupportsNamespaces, Configurable<Object> {
+
+  public static final String DEFAULT_BIGLAKE_SERVICE_ENDPOINT = "biglake.googleapis.com:443";
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigLakeCatalog.class);
+
+  // The name of this Iceberg catalog plugin: spark.sql.catalog.<catalog_plugin>.
+  private String name;
+  private Map<String, String> bigLakeProperties;
+  private FileIO io;
+  private Object conf;
+
+  private String bigLakeProjectId;
+  private String bigLakeRegion;
+  // BLMS catalog ID and fully qualified name.
+  private String catalogId;
+  private CatalogName catalogName;
+  private BigLakeClient bigLakeClient;
+
+  private CloseableGroup closeableGroup;
+
+  // Must have a no-arg constructor to be dynamically loaded
+  // initialize(String name, Map<String, String> properties) will be called to complete
+  // initialization
+  public BigLakeCatalog() {}
+
+  @Override
+  public void initialize(String inputName, Map<String, String> properties) {
+    Preconditions.checkArgument(
+        properties.containsKey(GCPProperties.BIGLAKE_PROJECT_ID),
+        "GCP project ID must be specified");
+    String projectId = properties.get(GCPProperties.BIGLAKE_PROJECT_ID);
+
+    Preconditions.checkArgument(
+        properties.containsKey(GCPProperties.BIGLAKE_GCP_REGION), "GCP region must be specified");
+    String region = properties.get(GCPProperties.BIGLAKE_GCP_REGION);
+
+    BigLakeClient client;
+    try {
+      // TODO: to add more auth options of the client. Currently it uses default auth
+      // (https://github.com/googleapis/google-cloud-java#application-default-credentials)
+      // that works on GCP services (e.g., GCE, GKE, Dataproc).
+      client =
+          new BigLakeClient(
+              properties.getOrDefault(
+                  GCPProperties.BIGLAKE_ENDPOINT, DEFAULT_BIGLAKE_SERVICE_ENDPOINT));
+    } catch (IOException e) {
+      throw new ServiceFailureException(e, "Creating BigLake client failed");
+    }
+
+    initialize(inputName, properties, projectId, region, client);
+  }
+
+  @VisibleForTesting
+  void initialize(
+      String inputName,
+      Map<String, String> properties,
+      String projectId,
+      String region,
+      BigLakeClient client) {
+    this.name = inputName;
+    this.bigLakeProperties = ImmutableMap.copyOf(properties);
+    this.bigLakeProjectId = projectId;
+    this.bigLakeRegion = region;
+    Preconditions.checkNotNull(client, "BigLake client must not be null");
+    this.bigLakeClient = client;
+
+    // Users can specify the BigLake catalog ID, otherwise catalog plugin name will be used.
+    // For example, "spark.sql.catalog.<plugin-name>=org.apache.iceberg.spark.SparkCatalog"
+    // specifies the plugin name "<plugin-name>".
+    this.catalogId =
+        this.bigLakeProperties.getOrDefault(GCPProperties.BIGLAKE_CATALOG_ID, inputName);
```

Comment:
Why is this pulled out of properties here, but the region and project ID are pulled out in the `initialize` method?

Reply:
My bad, region and project ID were used by constructing the BigLakeClient, but they are actually not used. I should move pulling region projectID to the other initialize method like catalog ID. Done.